download/doc/days2sec.3.man
NAME
   days2sec - [M_time] convert string of form [[-]dd-]hh:mm:ss.nn to seconds

SYNOPSIS
   function days2sec(str) result(time)

    character(len=*),intent(in)       :: str
    real(kind=realtime)               :: time

DESCRIPTION
     Given a string representing a duration of the form
     "[-][[[dd-]hh:]mm:]ss"  or NNdNNhNNmNNsNNw
     return a value representing seconds

     If "dd-" is present, units for the numbers are assumed to
     proceed from day to hour to minute to second. But if no
     day is present, the units are assumed to proceed from second
     to minutes to hour from left to right. That is ...

        [-]dd-hh:mm:ss
        [-]dd-hh:mm
        [-]dd-hh

        hh:mm:ss
        mm:ss
        ss

         Where dd is days, hh hours, mm minutes and ss seconds.  A decimal
         fraction is supported on the seconds.

         The numeric values may represent floating point numbers.  Spaces are ignored.

       NNdNNhNNmNNs
         Simple numeric values may also be used with unit suffixes; where
         s,m,h, or d represents seconds, minutes, hours or days and w
         represents a week. Allowed aliases for w,d,h,m, and s units are

                          d -  days,day
                          m -  minutes,minute,min
                          h -  hours,hour,hrs,hr
                          s -  seconds,second,sec
                          w -  week, weeks

         The numeric values may represent floating point numbers.

         Spaces, commas and case are ignored.

OPTIONS
      str   string of the general form dd-hh:mm:ss.nn
RETURNS
      time  the number of seconds represented by the input string

EXAMPLE
   Sample program:

    program demo_days2sec
    use M_time, only : days2sec
    implicit none
       write(*,*)days2sec('1-12:04:20')
       write(*,*)'one second ',days2sec('1')
       write(*,*)'one minute ',days2sec('1:00')
       write(*,*)'one hour ',days2sec('1:00:00')
       write(*,*)'one day ',days2sec('1-00:00:00')
       write(*,*)nint(days2sec(' 1-12:04:20              ')) .eq. 129860
       write(*,*)nint(days2sec(' 1.5 days                ')) .eq. 129600
       write(*,*)nint(days2sec(' 1.5 days 4hrs 30minutes ')) .eq. 145800
       write(*,*)nint(days2sec(' 1.5d                    ')) .eq. 129600
       write(*,*)nint(days2sec(' 1d2h3m4s                ')) .eq. 93784
       ! duplicates
       write(*,*)nint(days2sec(' 1d1d1d                  ')) .eq. 259200
       ! negative values
       write(*,*)nint(days2sec(' 4d-12h                  ')) .eq. 302400
    end program demo_days2sec

   Results:

    129860.00000000000
    one second    1.0000000000000000
    one minute    60.000000000000000
    one hour    3600.0000000000000
    one day    86400.000000000000
    T
    T
    T
    T
    T
    T
    T



download/tmp/PROGRAMS/days2sec.f90
subroutine help_usage(l_help)
implicit none
character(len=*),parameter     :: ident="@(#)help_usage(3f): prints help information"
logical,intent(in)             :: l_help
character(len=:),allocatable :: help_text(:)
integer                        :: i
logical                        :: stopit=.false.
stopit=.false.
if(l_help)then
help_text=[ CHARACTER(LEN=128) :: &
'NAME                                                                            ',&
'   days2sec - [TIME] Convert [[-]dd-][[hh:]mm:]ss to seconds                    ',&
'                                                                                ',&
'SYNOPSIS                                                                        ',&
'   days2sec dd-hh:mm:ss | --version| --help                                     ',&
'   days2sec NNdNNhNNmNNs                                                        ',&
'                                                                                ',&
'DESCRIPTION                                                                     ',&
'   Given a duration in the form dd-hh:mm:ss where dd is days, hh hours,         ',&
'   mm minutes and ss seconds convert it to seconds. Many utilities (ps(1),      ',&
'   for example) show times in this format to make it more intelligible;         ',&
'   but it generally easier to perform math on values represented in             ',&
'   seconds.                                                                     ',&
'                                                                                ',&
'OPTIONS                                                                         ',&
'   dd-hh:mm:ss  Given a string representing a duration of time in the           ',&
'                following forms:                                                ',&
'                                                                                ',&
'                  dd-hh:mm:ss                                                   ',&
'                     hh:mm:ss                                                   ',&
'                        mm:ss                                                   ',&
'                           ss                                                   ',&
'                                                                                ',&
'                convert it to seconds.                                          ',&
'                                                                                ',&
'                The numeric values may represent floating point numbers.        ',&
'                Spaces are ignored.                                             ',&
'                                                                                ',&
'    NNdNNhNNmNNs  Simple numeric values may also be used with unit suffixes;    ',&
'                  where s,m,h, or d represents seconds, minutes, hours          ',&
'                  or days. Allowed aliases for d,h,m, and s units are           ',&
'                                                                                ',&
'                   d -  days,day                                                ',&
'                   m -  minutes,minute,min                                      ',&
'                   h -  hours,hour,hrs,hr                                       ',&
'                   s -  seconds,second,sec                                      ',&
'                                                                                ',&
'                  The numeric values may represent floating point numbers.      ',&
'                  Spaces, commas  and case are ignored.                         ',&
'                                                                                ',&
'   --help       display this help and exit                                      ',&
'   --version    output version information and exit                             ',&
'                                                                                ',&
'EXAMPLE                                                                         ',&
'  Usage                                                                         ',&
'                                                                                ',&
'    days2sec 1-12:04:20                                                         ',&
'    129860                                                                      ',&
'    days2sec 1.5 days                                                           ',&
'    129600                                                                      ',&
'    days2sec 1.5 days 4hrs 30minutes                                            ',&
'    145800                                                                      ',&
'SEE ALSO                                                                        ',&
'    sec2days(1)                                                                 ',&
'']
   WRITE(*,'(a)')(trim(help_text(i)),i=1,size(help_text))
   stop ! if -help was specified, stop
endif
end subroutine help_usage
!-----------------------------------------------------------------------------------------------------------------------------------
!>
!!##NAME
!!    days2sec - [TIME] Convert [[-]dd-][[hh:]mm:]ss to seconds
!!
!!##SYNOPSIS
!!
!!    days2sec dd-hh:mm:ss | --version| --help
!!    days2sec NNdNNhNNmNNs
!!
!!##DESCRIPTION
!!    Given a duration in the form dd-hh:mm:ss where dd is days, hh hours,
!!    mm minutes and ss seconds convert it to seconds. Many utilities (ps(1),
!!    for example) show times in this format to make it more intelligible;
!!    but it generally easier to perform math on values represented in
!!    seconds.
!!
!!##OPTIONS
!!    dd-hh:mm:ss  Given a string representing a duration of time in the
!!                 following forms:
!!
!!                   dd-hh:mm:ss
!!                      hh:mm:ss
!!                         mm:ss
!!                            ss
!!
!!                 convert it to seconds.
!!
!!                 The numeric values may represent floating point numbers.
!!                 Spaces are ignored.
!!
!!     NNdNNhNNmNNs  Simple numeric values may also be used with unit suffixes;
!!                   where s,m,h, or d represents seconds, minutes, hours
!!                   or days. Allowed aliases for d,h,m, and s units are
!!
!!                    d -  days,day
!!                    m -  minutes,minute,min
!!                    h -  hours,hour,hrs,hr
!!                    s -  seconds,second,sec
!!
!!                   The numeric values may represent floating point numbers.
!!                   Spaces, commas  and case are ignored.
!!
!!    --help       display this help and exit
!!    --version    output version information and exit
!!
!!##EXAMPLE
!!
!!   Usage
!!
!!     days2sec 1-12:04:20
!!     129860
!!     days2sec 1.5 days
!!     129600
!!     days2sec 1.5 days 4hrs 30minutes
!!     145800
!!##SEE ALSO
!!     sec2days(1)
!===================================================================================================================================
subroutine help_version(l_version)
implicit none
character(len=*),parameter     :: ident="@(#)help_version(3f): prints version information"
logical,intent(in)             :: l_version
character(len=:),allocatable   :: help_text(:)
integer                        :: i
logical                        :: stopit=.false.
stopit=.false.
if(l_version)then
help_text=[ CHARACTER(LEN=128) :: &
'@(#)PRODUCT:        CLI library utilities and examples>',&
'@(#)PROGRAM:        days2sec(1f)>',&
'@(#)DESCRIPTION:    convert dd-hh:mm:ss string to seconds>',&
'@(#)VERSION:        1.0, 2016-06-17>',&
'@(#)AUTHOR:         John S. Urban>',&
'@(#)REPORTING BUGS: http://www.urbanjost.altervista.org/>',&
'@(#)HOME PAGE:      http://www.urbanjost.altervista.org/index.html>',&
'@(#)LICENSE:        Public Domain. This is free software: you are free to change and redistribute it.>',&
'@(#)                There is NO WARRANTY, to the extent permitted by law.>',&
'@(#)COMPILED:       Mon, Sep 11th, 2017 6:45:02 AM>',&
'']
   WRITE(*,'(a)')(trim(help_text(i)(5:len_trim(help_text(i))-1)),i=1,size(help_text))
   stop ! if -version was specified, stop
endif
end subroutine help_version
!-----------------------------------------------------------------------------------------------------------------------------------
program demo_days2sec
use M_kracken, only: kracken, sget, lget
use M_time,    only: days2sec
use M_strings, only: v2s
implicit none
character(len=*),parameter   :: ident="@(#) given string of form days-hh:mm:ss convert to seconds'"
character(len=:),allocatable :: printline
   call kracken('days2sec',' -oo -help .F. -version .F.')                    ! parse command line
   call help_usage(lget('days2sec_help'))                                    ! display help information and stop if true
   call help_version(lget('days2sec_version'))                               ! display version information and stop if true
   ! get value from command line with SGET, convert to seconds with DAYS2SEC, and make into a nicer value string with V2S
   printline=v2s(days2sec(sget('days2sec_oo'))) ! call this way instead of in write statement so error message will print
   write(*,'(a)')printline
end program demo_days2sec
